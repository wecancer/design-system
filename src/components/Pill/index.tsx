import React from 'react'
import styled, {css} from 'styled-components'
import {ColorsTemplate} from '../../styles/theme'

import Icon from '../Icon'

type Size = 'small' | 'medium' | 'large'

const Text = styled.p`
  ${({theme}) => css`
    font-family: ${theme.font.familyRedesign};
    font-weight: ${theme.font.weights.regular};
    margin: 0;
  `}
`

const Container = styled.div<
  {fillColor: ColorsTemplate} & Pick<Props, 'size' | 'isActive'>
>`
  ${({isActive, theme, size, fillColor}) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    white-space: nowrap;
    border: 1px solid ${theme.colors[fillColor]};

    &[role='button'] {
      cursor: pointer;
    }

    ${
      isActive
        ? css`
            color: ${theme.colors.offWhite};
            background-color: ${theme.colors[fillColor]};
          `
        : css`
            color: ${theme.colors[fillColor]};
            background-color: ${theme.colors.offWhite};
          `
    }}

    ${(() => {
      if (size === 'large') {
        return css`
          height: 2.875rem;
          font-size: 1rem;
          border-radius: 1.5rem;
          padding: 0 1rem;
        `
      }
      if (size === 'small') {
        return css`
          height: 1.875rem;
          font-size: 0.8125rem;
          border-radius: 1rem;
          padding: 0 0.75rem;
        `
      }
      return css`
        height: 2.375rem;
        font-size: 0.875rem;
        border-radius: 1.25rem;
        padding: 0 0.75rem;
      `
    })()}
  `}
`

type IconProps = {
  isActive: boolean
  fillColor: ColorsTemplate
}

const IconStyled = styled(Icon)<IconProps>`
  ${({isActive, theme, fillColor}) =>
    isActive
      ? css`
          color: ${theme.colors.offWhite};
        `
      : css`
          color: ${theme.colors[fillColor]};
        `}
`

const CheckIcon = styled(IconStyled)`
  margin-right: 0.5rem;
`

const CloseIcon = styled(IconStyled)`
  margin-left: 0.5rem;
`

export type Props = {
  size?: Size
  label: string
  isActive?: boolean
  hasClose?: boolean
  isChecked?: boolean
  onClick?: () => void
  fillColor?: ColorsTemplate
}

const Pill = ({
  label,
  onClick,
  size = 'medium',
  isActive = false,
  hasClose = false,
  fillColor = 'primary',
  isChecked = false,
}: Props) => {
  const isClickable = !!onClick
  const checkIconSize = size === 'large' ? '16px' : '14px'
  const closeIconSize = size === 'large' ? '20px' : '16px'

  const clickProps = {
    role: 'button',
    onClick,
    onKeyDown: onClick,
  }

  return (
    <Container
      size={size}
      isActive={isActive}
      fillColor={fillColor}
      {...(isClickable ? clickProps : {})}
    >
      {isChecked && (
        <CheckIcon
          name="check"
          isActive={isActive}
          fillColor={fillColor}
          size={checkIconSize}
        />
      )}
      <Text>{label}</Text>
      {hasClose && (
        <CloseIcon
          name="close"
          isActive={isActive}
          fillColor={fillColor}
          size={closeIconSize}
        />
      )}
    </Container>
  )
}

export default Pill
