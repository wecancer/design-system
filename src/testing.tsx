import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
