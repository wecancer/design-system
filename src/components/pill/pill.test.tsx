import Pill from '.'
import { render, fireEvent } from '../../testing'

describe('Pill Component', () => {
  it('should render basic Pill Component correctly', () => {
    const { container } = render(
      <>
        <Pill label="test" />
        <Pill label="test actived" isActive />
      </>,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render icons of Pill Component correctly', () => {
    const { container } = render(
      <>
        <Pill label="has closed" hasClose />
        <Pill label="checked" isChecked />
        <Pill label="checked and has closed" hasClose isChecked />
      </>,
    )
    expect(container).toMatchSnapshot()
  })

  it('should render checked and activied Pill Component correctly', () => {
    let item = 'none'
    const { container } = render(
      <Pill
        label="test"
        onClick={() => {
          item = 'clicked'
        }}
      />,
    )
    fireEvent.click(container.firstChild as HTMLElement)
    expect(item).toBe('clicked')
  })
})
