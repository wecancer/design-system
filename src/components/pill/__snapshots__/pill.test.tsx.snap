// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pill Component should render basic Pill Component correctly 1`] = `
<div>
  <div
    class="sc-dkPtRN chlKdc"
  >
    <p
      class="sc-gsDKAQ eHDJig"
    >
      test
    </p>
  </div>
  <div
    class="sc-dkPtRN cVnAYw"
  >
    <p
      class="sc-gsDKAQ eHDJig"
    >
      test actived
    </p>
  </div>
</div>
`;

exports[`Pill Component should render icons of Pill Component correctly 1`] = `
<div>
  <div
    class="sc-dkPtRN chlKdc"
  >
    <p
      class="sc-gsDKAQ eHDJig"
    >
      has closed
    </p>
    <div
      class="sc-bdvvtL hZeSqt sc-hKwDye sc-jRQBWg bhCFhM fmYyGE"
      data-icon-name="close"
    >
      <svg
        fill="none"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 6L18.7742 18.7742"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
        <path
          d="M6 18.7742L18.7742 5.99998"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
  <div
    class="sc-dkPtRN chlKdc"
  >
    <div
      class="sc-bdvvtL iEQYcl sc-hKwDye sc-eCImPb bhCFhM hWCMKE"
      data-icon-name="check"
    >
      <svg
        fill="none"
        viewBox="0 0 18 13"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17 1L6 12L1 7"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <p
      class="sc-gsDKAQ eHDJig"
    >
      checked
    </p>
  </div>
  <div
    class="sc-dkPtRN chlKdc"
  >
    <div
      class="sc-bdvvtL iEQYcl sc-hKwDye sc-eCImPb bhCFhM hWCMKE"
      data-icon-name="check"
    >
      <svg
        fill="none"
        viewBox="0 0 18 13"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17 1L6 12L1 7"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <p
      class="sc-gsDKAQ eHDJig"
    >
      checked and has closed
    </p>
    <div
      class="sc-bdvvtL hZeSqt sc-hKwDye sc-jRQBWg bhCFhM fmYyGE"
      data-icon-name="close"
    >
      <svg
        fill="none"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 6L18.7742 18.7742"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
        <path
          d="M6 18.7742L18.7742 5.99998"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
</div>
`;
